src/lib/optimizer/join_ordering/abstract_join_ordering_algorithm.cpp

```diff
@@ -80,30 +78,32 @@ std::shared_ptr<AbstractLQPNode> AbstractJoinOrderingAlgorithm::_add_join_to_pla
   std::sort(join_predicates_and_cost.begin(), join_predicates_and_cost.end(),
             [&](const auto& lhs, const auto& rhs) { return lhs.second < rhs.second; });
 
-  // Find the simple predicate with the lowest cost (if any exists), which will act as the primary predicate
-  auto primary_join_predicate = std::shared_ptr<AbstractExpression>{};
-  for (auto predicate_iter = join_predicates_and_cost.begin(); predicate_iter != join_predicates_and_cost.end();
-       ++predicate_iter) {
-    // If a predicate can be converted into an OperatorJoinPredicate, it can be used as a primary predicate
-    const auto operator_join_predicate =
-        OperatorJoinPredicate::from_expression(*predicate_iter->first, *left_lqp, *right_lqp);
+  // Categorize join predicates into those that can be processed as part of a join operator and those that need to be
+  // processed as scans.
```

Comment:
One more question: Is every predicate that can be a secondary predicate now handled as such? What are the performance implications of this? Are there still cases where a dedicated scan would be more efficient?

Reply:
I have just adjusted JoinOrderingRule such that NO secondary predicates will be part of JoinOperators - as scans are probably still faster. I am running TPCH and JOB as we speak and will update the OP with the results. I expect a slight decline in perf due to the checks for MPJ presence for each tuple that passes the primary predicate - but it shouldn't be too much. Let's see.

Reply:
Also added to OP: https://gist.github.com/mrzzzrm/b55d39a1214acf9208a473e5d222a8bc

All is well, I'd say.

Reply:
> such that NO secondary predicates will be part of JoinOperators

Unless it is syntactically necessary, correct?

Reply:
Not sure if you have seen this. This would be my last concern for this PR.

Reply:
The JoinOrderingRule is currently only concerned with inner joins, so emitting secondary predicates would never be "necessary" for it.

This PR does NOT enable multiple join predicates for outer joins (#1436), but this is probably just the case of removing an Assert in the SQLTranslator and adding tests.